fix(#596) fix issue with semver validation for latest versions

1yuv opened the pull request:
# Description

[description]

medic/cht-conf#596
# Code review items

- Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- Tested: Unit and/or integration tests where appropriate
- Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

kennsippell commented on src/fn/upload-app-settings.js:
```diff
@@ -39,7 +39,7 @@ async function augmentDeclarativeWithNoolsBoilerplate(appSettings) {
 
   const actualCoreVersion = await getValidApiVersion(appSettings);
-  const addNoolsBoilerplate = semver.lt(actualCoreVersion, '4.2.0-dev');
+  const addNoolsBoilerplate = actualCoreVersion && semver.lt(actualCoreVersion, '4.2.0-dev');
```
I'm unclear on when `actualCoreVersion` would be falsy... Why?

1yuv replied:
Even though coerce tries to parse versions correctly, it'd be falsy when a custom version like tag number or branch number is being used on production. What do you suggest in this case to avoid scenarios like you mentioned? One option is to use `base_Version` which is only available in latest CHT versions and ensure this doesn't get changed. This issue was also caused because the way version was written was changed.

kennsippell replied:
Can you share the specific tags you're having trouble with please? I'm testing with coerce and not clear on what tags are failing.

1yuv replied:
This issue fix was needed for regular CHT versions as well like `4.5.0.6922454971`, hence I introduced coerce for that fix. `actualCoreVersion` could still be falsy if production instances are running on versions like [this](https://github.com/medic/config-nssd/issues/65) (private). We won't need falsy check after introducing coerce. I don't see recent instances with unusual versioning though. I am not sure what `coreVersion` would've returned for particular versions like that since `api/deploy-info` endpoint has changed frequently.

kennsippell commented on src/lib/get-api-version.js:
```diff
@@ -18,7 +18,7 @@ const getApiVersion = async () => {
 
 const getValidApiVersion = async () => {
-  return semver.valid(await getApiVersion());
+  return semver.valid(semver.coerce(await getApiVersion()));
```
Isn't the coerce all that would be needed for the `4.5.0.6922454971` example you listed in https://github.com/medic/cht-conf/issues/596?

1yuv replied:
`getValidApiVersion()` is being used elsewhere as well and valid and coerce return different types.

kennsippell reviewed (changes requested):
Can you please add a test to capture this scenario clearly?

kennsippell commented on test/fn/upload-app-settings.spec.js:
```diff
@@ -20,6 +20,10 @@ describe('upload-app-settings', () => {
     // case when testing on dev
     { coreVersion: '4.2.0-dev.1682192676689', isDeclarative: true, expectNools: false },
+    // https://github.com/medic/cht-conf/issues/596
+    { coreVersion: '4.5.0.6922454971', isDeclarative: false, expectNools: false},
```
shouldn't this be `isDeclarative: true`?

1yuv commented on test/fn/upload-app-settings.spec.js:
```diff
@@ -20,6 +20,10 @@ describe('upload-app-settings', () => {
     // case when testing on dev
     { coreVersion: '4.2.0-dev.1682192676689', isDeclarative: true, expectNools: false },
+    // https://github.com/medic/cht-conf/issues/596
+    { coreVersion: '4.5.0.6922454971', isDeclarative: false, expectNools: false},
```
It was failing with version `4.5.0.6922454971`, if we pass weird build version, it'd faill.

kennsippell replied:
If I revert your changes in `src`, all these tests are still passing... This means that you have not sufficiently captured the scenario you care about in these tests. Am I mistaken?

1yuv replied:
That's right, because the fix was on `getValidApiVersion()` and in this example, we're mocking what getValidApiVersion() returns. Previously it was failing on getValidApiVersion() and failing `upload-app-settings`. I can add unit test for `getApiVersion()` method if that makes it more useful.

kennsippell replied:
This isn't a super obvious requirement, so if you want it to not break in the future - it is typically nice to have a test which captures the requirement. Ya - that can be a unit test for `getApiVersion` or an integration test here (change what is mocked to be deeper).

jkuester reviewed (approved):
I have nothing to add on top of what Kenn has said!  Approving this preemptively so you can merge as soon as Kenn's concerns are resolved!

1yuv commented:
Hi @kennsippell , I've added few more scenarios for unit test. The original inetnt of this issue was to fix the scenarios where valid CHT core versions like 4.5.0.xxxx were not being parsed properly and an error was being thrown.

Before this fix, if you were running `upload-app-settings` against any CHT instance with 4.5.0, it'd fail because `/api/deployinfo` returns version `4.5.0.6922454971` causing `actualCoreVersion` to be falsy and crash when running `semver.lt()`. semver.coerce would correctly parse 4.5.0.6922454971 to `4.5.0` and fix the problem. 

Even after adding coerce, I added check on if `actualCoreVersion` is undefined or not to stop it from failing. 

 ```js
const addNoolsBoilerplate = actualCoreVersion && semver.lt(actualCoreVersion, '4.2.0-dev');
```

However, if you think we should still crash the app and prevent users from running `cht upload-app-settings` in non-standard versions, I can remove that undefined check. 

If you're looking for previous isntances where CHT version was not a valid semver, you can see these on [this](https://docs.google.com/spreadsheets/d/1H6ar1yIAvWDKcsNzvxmi9GhZ-0oQAwNBMYA66XAUHG0/edit#gid=1200196719) (pvt) scraper output.

kennsippell reviewed (approved):
Leave it to you to decide whether to add the last automated tests.

1yuv commented:
Oops, missed a colon again while merging, sending another PR to release @kennsippell , @jkuester
